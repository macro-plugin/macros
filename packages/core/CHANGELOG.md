# @macro-plugin/core

## 1.2.0

### Minor Changes

- [#33](https://github.com/macro-plugin/macros/pull/33) [`8c94045`](https://github.com/macro-plugin/macros/commit/8c9404545d207b708ed433328bd383488a3ed0d4) Thanks [@voorjaar](https://github.com/voorjaar)! - feat: new cursor context method

- [#33](https://github.com/macro-plugin/macros/pull/33) [`b45063f`](https://github.com/macro-plugin/macros/commit/b45063ff784e513380f43bf8425c4f72eaa5c0b4) Thanks [@voorjaar](https://github.com/voorjaar)! - refactor: rename span to dummySpan

## 1.1.4

### Patch Changes

- [#29](https://github.com/macro-plugin/macros/pull/29) [`e3ce376`](https://github.com/macro-plugin/macros/commit/e3ce37696e0a21bdddb56804f7dbcfa904f5fd35) Thanks [@voorjaar](https://github.com/voorjaar)! - feat: support resolveTs and swcTransform options

## 1.1.3

## 1.1.2

### Patch Changes

- [#23](https://github.com/macro-plugin/macros/pull/23) [`294f675`](https://github.com/macro-plugin/macros/commit/294f67574c5ff168bdeb0a499ea4ab2fd57579cf) Thanks [@voorjaar](https://github.com/voorjaar)! - feat: support include, exclude, exclude, extensions, tsconfig options

## 1.1.1

### Patch Changes

- [#21](https://github.com/macro-plugin/macros/pull/21) [`b4be79b`](https://github.com/macro-plugin/macros/commit/b4be79be09e8790b95c6a2e9b07dde4ba3822f39) Thanks [@voorjaar](https://github.com/voorjaar)! - chore: update to swc 1.3.55

- [#21](https://github.com/macro-plugin/macros/pull/21) [`df8dcc0`](https://github.com/macro-plugin/macros/commit/df8dcc0ab7f3e09f1157c5c2a7cda4a7f5367991) Thanks [@voorjaar](https://github.com/voorjaar)! - fix: use swc Plugin type exports

## 1.1.0

### Minor Changes

- [#16](https://github.com/macro-plugin/macros/pull/16) [`4155bb4`](https://github.com/macro-plugin/macros/commit/4155bb4de7968a83e62203411bceae6b0602637f) Thanks [@voorjaar](https://github.com/voorjaar)! - feat: support loading external macros

### Patch Changes

- [#16](https://github.com/macro-plugin/macros/pull/16) [`6b25910`](https://github.com/macro-plugin/macros/commit/6b25910567e910b7c71c79646f8569a2f3927be6) Thanks [@voorjaar](https://github.com/voorjaar)! - fix: types for $Quote template macro

- [#16](https://github.com/macro-plugin/macros/pull/16) [`8dd4939`](https://github.com/macro-plugin/macros/commit/8dd493997931d8d91a82ffb1785927d425c17c61) Thanks [@voorjaar](https://github.com/voorjaar)! - fix: tracker failed to load expression macro when using import

- [#16](https://github.com/macro-plugin/macros/pull/16) [`14efcb2`](https://github.com/macro-plugin/macros/commit/14efcb2c6461ab3f5d78e0599ec74b422085ce1d) Thanks [@voorjaar](https://github.com/voorjaar)! - feat: support access config from macro context

- [#16](https://github.com/macro-plugin/macros/pull/16) [`7787022`](https://github.com/macro-plugin/macros/commit/7787022f657b7a79cb18a1e8ba947ae2eaeb682e) Thanks [@voorjaar](https://github.com/voorjaar)! - fix: mark proxy as macro_plugin

- [#16](https://github.com/macro-plugin/macros/pull/16) [`a17b055`](https://github.com/macro-plugin/macros/commit/a17b055d356c285b394add56192c80077ebde2c2) Thanks [@voorjaar](https://github.com/voorjaar)! - feat: new AST types exported from swc

## 1.0.3

### Patch Changes

- [#12](https://github.com/macro-plugin/macros/pull/12) [`07961a0`](https://github.com/macro-plugin/macros/commit/07961a03b6e82080a2b8c8ab2626c187c34f912e) Thanks [@voorjaar](https://github.com/voorjaar)! - refactor: move createLabeledExpr and createLabeledBlock to labeled

- [#12](https://github.com/macro-plugin/macros/pull/12) [`f1e40ea`](https://github.com/macro-plugin/macros/commit/f1e40ead32636d4f2d43c442e70cc208e2d43b28) Thanks [@voorjaar](https://github.com/voorjaar)! - fix: decorator in arrowFunction with variableDeclarator should keep name

- [#12](https://github.com/macro-plugin/macros/pull/12) [`556ca2d`](https://github.com/macro-plugin/macros/commit/556ca2d9addaf36ac84da8c8ea7b5bc465e174b7) Thanks [@voorjaar](https://github.com/voorjaar)! - feat: support labeled function decorator

## 1.0.2

### Patch Changes

- [#9](https://github.com/macro-plugin/macros/pull/9) [`2c64787`](https://github.com/macro-plugin/macros/commit/2c647875182c5fc6ca41c9e72587a08307ba90ec) Thanks [@voorjaar](https://github.com/voorjaar)! - chore: upgrade to swc 1.3.53

## 1.0.1

### Patch Changes

- 98425cf: build: no longer publish iife
