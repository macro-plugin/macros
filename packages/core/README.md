# @macro-plugin/core

Macro system for JavaScript and TypeScript.

## Install

```sh
# if you use npm
npm i -D @macro-plugin/core
# if you use pnpm
pnpm i -D @macro-plugin/core
# if you use yarn
yarn add -D @macro-plugin/core
```

## Why

Have you ever used Rust? Rust has an impressive macro system. I'm wondering why the JavaScript world can't have one. Now you got it.

## How it works

It uses [swc](https://swc.rs/)'s parser and printer, and does not make any conversions other than macors to your project by default. Of course, it can also be used as a swc plugin, so we can transform typescript and jsx.

## Features

- A new walker for macros, inspired by [estree-walker](https://github.com/Rich-Harris/estree-walker)
- Built-in variable tracker plugin, which makes it easier to convert variables.
- Very simple yet powerful APIs for creating various types of macros.
- Factory runtime and macros for simplify creating swc AST.
- Support automatic generation of macro declarations file. (macros.d.ts)
- Many common built-in macros.
- Support all web framework intergrations.

## Built-in macros

The project is still in the beta stage, and there are more macros under development.

### Expression Macros

Expression macros may be a very familiar macro type. It converts a `CallExpression` into any other Expression.

#### `$Eval`

compile time eval.

#### `$Ast`

convert anything into an ast.

#### `$Env`

read env and replace with result.

#### `$Stringify`

convert any expression or type into string

#### `$Span`

get a span like `[line, column]`

#### `$Line`

get current line number

#### `$Column`

get current column number

#### `$ID`

generate unique id based on `[line, column, filename]`

#### `$Include`

include another `js, ts` file into current script.

#### `$IncludeStr`

include a string from any path

#### `$IncludeJSON`

include a json, also support include a json value with key.

#### `$WriteFile`

compile time write to file

#### `$Concat`

compile time concat string

#### `$Expr`

turn code into an expression

#### `$Quote`

a template macro for create macro

#### `$UnImplemented`

mark code as unImplemented, log error when compile, throw error in client side.

#### `$Todo`

mark code as todo, log warning when compile, throw error in client side.

#### `$UnReachable`

mark code as unreachable, throw error in client side.

...

### Labeled Macros

LabeledMacros are very interesting feature. It even allows you to invent your own grammar.
For example:

```js
debug: {
  console.log("something")
}
```

or creating states:

```js
state: {
  var count = 0
  var name = "macro"
}
```

or using function decorator:

```js
function myFunc() {
  decorator: [yourFunctionDecorator]

  doSomething()
}
```

you can even create macro with it.

```ts
macro: {
  var __DEV__ = true
  
  var add = $ExprMacro<(a: number, b: number) => number>(function(args) {
    const a = this.printExpr(args[0])
    const b = this.printExpr(args[1])

    if (+a < 0) return args[1]
    return this.parseExpr(`${a} + ${b}`)
  })
}

if (__DEV__) {
  const a = add(1, 2)
}
```

transform to

```js
if (true) {
  const a = 1 + 2
}
```


To use Labeled Macros, you may need some extra config for typescript and eslint.

#### Typescript

```json
{
  "compilerOptions": {
    // ...
    "allowUnusedLabels": true,
  },
}
```

#### Eslint

```js
module.exports = {
  rules: {
    // ...
    "no-labels": 0,
  }
}
```

## API

### `transform/transformAst/transformAsync`

transform code with macro plugins.

### `createSwcPlugin`

use as swc plugin.

### `walk`

walk an ast node or node list

### `print/printAsync/printExpr/printType`

print ast to code

### `parse/parseAsync/parseExpr/parseType`

parse code to ast

### `defineConfig`

define config with types

### `createMacro`

create a macro

### `createLitMacro`

create a literal macro

### `createExprMacro`

create an expression macro

### `createTypeMacro`

create a type macro

### `createTmplMacro`

create a template macro

### `createLabeledMacro`

create a labeled macro

### `isMacroPlugin`

check if input is a macro plugin

## FAQ

### LabeldMacros confusing with JavaScript

Some people may argue about labeldMacros, confusing the original semantics of JavaScript. But in fact, `labeledStatement` and `var` are both forgotten or discarded syntax of JavaScript. So we bring something new to it. Personally, I don't think this is a bad thing.

### Compare with babel

In fact, the beginning of this project based on babel's parser and generator. When I switched to swc, I did a simple test and found that the speed was at least tripled, so it must be faster.

This project is based on the AST type of swc and is not compatible with Babel's AST. And the project focuses on transforming macros. If you need the support of something like browserlist, you need to use it with swc.

### Compare with swc

You can regard it as an enhanced version of swc with macro support, as well as better aapi. However, we do not use the `.swcrc` configuration file. The json configuration is not friendly for macros, and you should define all your swc configuration in `macros.config.js` or `macros.config.ts`.

## License

[MIT](https://github.com/macro-plugin/macros/blob/main/LICENSE)

Copyright (c) 2023, Raven Satir