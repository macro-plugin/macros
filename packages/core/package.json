{
  "name": "@macro-plugin/core",
  "version": "1.0.1",
  "description": "Macro system for JavaScript and TypeScript.",
  "main": "dist/index.js",
  "module": "dist/index.mjs",
  "types": "./dist/index.d.ts",
  "bin": {
    "macros": "dist/cli.js"
  },
  "files": [
    "dist",
    "CHANGELOG.md"
  ],
  "keywords": [
    "macro",
    "macros",
    "cli",
    "macro-plugin",
    "macro-system"
  ],
  "scripts": {
    "build": "rollup -c ../../rollup.config.mjs"
  },
  "dependencies": {
    "@swc/core": "^1.3.40"
  },
  "homepage": "https://github.com/macro-plugin/macros/tree/main/packages/core",
  "repository": {
    "type": "git",
    "url": "https://github.com/macro-plugin/macros.git"
  },
  "sideEffects": false,
  "exports": {
    "./package.json": "./package.json",
    ".": {
      "import": "./dist/index.mjs",
      "require": "./dist/index.js",
      "types": "./dist/index.d.ts"
    },
    "./cli": {
      "require": "./dist/cli.js"
    },
    "./register": {
      "require": "./dist/register.js"
    }
  },
  "author": "Raven Satir",
  "license": "MIT"
}
