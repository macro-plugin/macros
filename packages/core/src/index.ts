export { createLabeledBlock } from "./block"
export { createLabeledExpr } from "./expr"
export { createSwcPlugin, transform, transformAst, transformAsync } from "./transform"
export { print, printAsync, printExpr, printExprAsync } from "./print"
export { parse, parseAsync, parseExpr, parseType } from "./parse"
export { walk, Walker } from "./walk"
export { macro, $Macro, $LitMacro, $ExprMacro, $TmplMacro, $TypeMacro, $LabeledMacro } from "./macro"
export * from "./api"
export * from "./lib"
export * from "./types"
export * from "./utils"
