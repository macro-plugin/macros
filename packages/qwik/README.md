# Qwik Macros

Macros for [qwik](https://qwik.builder.io/).

## qwik

Macro for create a qwik component.

```jsx
export const HelloWorld = () => {
  qwik: true

  return <div>Hello world</div>
};
```

or 

```jsx
export function HelloWorld() {
  qwik: true

  return <div>Hello world</div>
}
```

compiles to

```jsx
import { $component } from "@builder.io/qwik"

export const HelloWorld = $component(() => {
  return <div>Hello world</div>
})

```

## signal

Macro for using qwik signals.

```jsx
export function Counter() {
  qwik: true
  signal: {
    var count = 0
  }
  return (
    <>
      <div>Value is: {count}</div>
      <button onClick$={() => count++}>Increment</button>
    </>
  );
}
```

compiles to

```jsx
import { useSignal, $component } from "@builder.io/qwik"

export const Counter = component$(() => {
  const count = useSignal(0);
  return (
    <>
      <div>Value is: {count.value}</div>
      <button onClick$={() => count.value++}>Increment</button>
    </>
  );
});
```

## computed

Macro for create a reactive variable.

```jsx
function ComputedExample() {
  qwik: true
  signal: {
    var count = 1
  }
  computed: {
    var doubleCount = 2 * count
  }
  return <div>{count} / {doubleCount}</div>
}
```

compiles to

```jsx
import { useSignal } from "@builder.io/qwik";
import { useTask$ } from "@builder.io/qwik";
import { $component } from "@builder.io/qwik"

const ComputedExample = component$(() => {
  const count = useSignal(1);
  const doubleCount = useSignal(2 * count.value);
  useTask$(({ track  })=>{
    const __count = track(()=>count.value);
    doubleCount.value = 2 * __count;
  });
  return <div>{count} / {doubleCount}</div>
})
```

## store

Macro for using qwik store.

```jsx
export function LocalStateExample() {
  qwik: true
  store: {
    var state = {
      value: 0
    }
  }
  return <div>Value is: {state.value}</div>
}
```

compiles to

```jsx
import { useStore, $component } from "@builder.io/qwik"

export const LocalStateExample = component$(() => {
  const state = useStore({
    value: 0,
  });
  return <div>Value is: {state.value}</div>;
});
```

## resource

Macro for qwik `useResource$`.

```js
function ResourceExample() {
  qwik: true
  signal: {
    var bar = 'foo'
  }
  resource: async (ctx) => {
    ctx.track(() => bar);
    ctx.track(() => props.url);
    ctx.cleanup(() => console.log('cleanup'));

    var someResource = await expensiveCompute(bar, props.url);
  }
  return <div></div>
}
```

compiles to

```js
import { useSignal } from "@builder.io/qwik";
import { useResource$ } from "@builder.io/qwik";
import { component$ } from "@builder.io/qwik";

const ResourceExample = component$(() => {
  const bar = useSignal("foo");
  const someResource = useResource$(async (ctx)=>{
      ctx.track(()=>bar.value);
      ctx.track(()=>props.url);
      ctx.cleanup(()=>console.log("cleanup"));
      return await expensiveCompute(bar.value, props.url);
  });
})
```

## task

Macro for qwik `useTask$`.

```js
function DoubleCounter() {
  qwik: true
  signal: {
    var count = 1
    var doubleCount = 0
  }
  task: {
    console.log('component mounted')
  }
  task: ({track}) => {
    const newCount = track(() => count)
    doubleCount = 2 * newCount
  }
  return <div>{count} / {doubleCount}</div>
}
```

compiles to

```js
import { useSignal } from "@builder.io/qwik";
import { useTask$ } from "@builder.io/qwik";
import { component$ } from "@builder.io/qwik";

const DoubleCounter = component$(() => {
  const count = useSignal(1);
  const doubleCount = useSignal(0);
  useTask$(()=>{
    console.log("component mounted");
  });
  useTask$(({ track  })=>{
    const newCount = track(()=>count.value);
    doubleCount.value = 2 * newCount;
  });
  return <div>{count} / {doubleCount}</div>;
})
```

## vtask

Macro for qwik `useVisibleTask$`

```js
export function Clock() {
  qwik: true
  signal: {
    var seconds = 0
  }
  vtask: {
    const interval = setInterval(() => {
      seconds++
    }, 1000)
    return () => clearInterval(interval)
  }
  return <div>Seconds: {seconds}</div>
}
```

compiles to

```js
import { component$, useSignal, useVisibleTask$ } from "@builder.io/qwik"

export const Clock = component$(() => {
  const seconds = useSignal(0);
  useVisibleTask$(() => {
    const interval = setInterval(() => {
      seconds.value++;
    }, 1000);
    return () => clearInterval(interval);
  });

  return <div>Seconds: {seconds.value}</div>;
});
```

## events

Apply for all labels that starts with `on`, add event listener for qwik component.

```js
export function ClickableComponent() {
  qwik: true
  onclick: {
    alert('Alert from Clickable Component.');
  }

  return <div>click from other component 1</div>;
}
```

compiles to

```js
import { component$ } from "@builder.io/qwik";
import { $ } from "@builder.io/qwik";
import { useOn } from "@builder.io/qwik";
export const ClickableComponent = component$(()=>{
  useOn("click", $(()=>{
    alert("Alert from Clickable Component.");
  }));
  return <div>click from other component 1</div>;
});
```

### document

```js
function KeyBoard() {
  qwik: true
  signal: {
    var keyPressed = ''
  }
  onkeydown: (event) => {
    document: true
    keyPressed = keyPressed + event.key;
  }

  return <div>{keyPressed}</div>;
}
```

or

```js
function KeyBoard() {
  qwik: true
  signal: {
    var keyPressed = ''
  }
  document: {
    onkeydown: (event) => {
      keyPressed = keyPressed + event.key;
    }
    onkeyup: (event) => {
      console.log('keyup')
    }
  }

  return <div>{keyPressed}</div>;
}
```

compiles to

```js
import { component$ } from "@builder.io/qwik";
import { useSignal } from "@builder.io/qwik";
import { $ } from "@builder.io/qwik";
import { useOnDocument } from "@builder.io/qwik";
const KeyBoard = component$(()=>{
  const keyPressed = useSignal("");
  useOnDocument("keydown", $((event)=>{
    keyPressed.value = keyPressed.value + event.key;
  }));
  return <div>{keyPressed.value}</div>;
});
```

### window

```js
export function Online() {
  qwik: true
  window: {
    ononline: {
      alert('Your Device is now Online')
    }
    onoffline: {
      alert('Your Device is now Offline')
    }
  }

  return <div></div>
};
```

or

```js
export function Online() {
  qwik: true
  ononline: {
    window: true
    alert('Your Device is now Online')
  }
  onoffline: {
    window: true
    alert('Your Device is now Offline')
  }

  return <div></div>
}
```

compiles to

```js
import { component$ } from "@builder.io/qwik";
import { $ } from "@builder.io/qwik";
import { useOnWindow } from "@builder.io/qwik";

export const Online = component$(()=>{
  useOnWindow("online", $(()=>{
    alert("Your Device is now Online");
  }));
  useOnWindow("offline", $(()=>{
    alert("Your Device is now Offline");
  }));
  return <div></div>;
})
```

## link

Macro for import a css file.

### single

```js
export const CmpStyles = () => {
  qwik: true
  link: './code-block.css?inline'

  return <span class="my-text">Some text</span>
}
```

compiles to

```js
import { component$ } from "@builder.io/qwik";
import { useStyles$ } from "@builder.io/qwik";
import __link0 from "./code-block.css?inline";

export const CmpStyles = component$(()=>{
  useStyles$(__link0);
  return <span class="my-text">Some text</span>;
});
```

### multiple

```js
export const CmpStyles = () => {
  qwik: true
  link: [
    './code-block.css?inline',
    './font-style.css?inline',
  ]

  return <span class="my-text">Some text</span>
}
```

compiles to

```js
import { component$ } from "@builder.io/qwik";
import { useStyles$ } from "@builder.io/qwik";
import __link0 from "./code-block.css?inline";
import __link1 from "./font-style.css?inline";

export const CmpStyles = component$(()=>{
  useStyles$(__link0);
  useStyles$(__link1);
  return <span class="my-text">Some text</span>;
});
```

## css

Macro for use inline css.

```js
export const CmpStyles = () => {
  qwik: true
  css: `
    .my-text {
      color: red;
    }
  `
  return <span class="my-text">Some text</span>
}
```

compiles to

```js
import { component$ } from "@builder.io/qwik";
import { useStyle$ } from "@builder.io/qwik";

export const CmpStyles = component$(()=>{
  useStyle$(`
    .my-text {
      color: red;
    }
  `);
  return <span class="my-text">Some text</span>;
});
```

## scoped

Macro for use scoped css link or string.

```js
export const CmpStyles = () => {
  qwik: true
  scoped: {
    link: './code-block.css?inline'

    css: `
      .my-text {
        color: red;
      }
    `
  }

  return <span class="my-text">Some text</span>
}
```

compiles to

```js
import { component$ } from "@builder.io/qwik";
import { useStyleScoped$ } from "@builder.io/qwik";
import __link0 from "./code-block.css?inline";

export const CmpStyles = component$(()=>{
  useStyleScoped$(__link0);
  useStyleScoped$(`
    .my-text {
      color: red;
    }
  `);
  return <span class="my-text">Some text</span>;
});
```