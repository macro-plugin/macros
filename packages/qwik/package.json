{
  "name": "@macro-plugin/qwik",
  "version": "1.0.0",
  "description": "A simpler way of creating components for qwik by using macros",
  "main": "dist/index.js",
  "module": "dist/index.mjs",
  "types": "dist/index.d.ts",
  "files": [
    "dist",
    "CHANGELOG.md"
  ],
  "keywords": [
    "qwik",
    "qwikjs",
    "macro",
    "macros",
    "macro-plugin"
  ],
  "scripts": {
    "build": "rollup -c ../../rollup.config.mjs"
  },
  "homepage": "https://github.com/macro-plugin/macros/tree/main/packages/qwik",
  "repository": {
    "type": "git",
    "url": "https://github.com/macro-plugin/macros.git"
  },
  "dependencies": {
    "@macro-plugin/core": "1.0.0"
  },
  "exports": {
    "./package.json": "./package.json",
    ".": {
      "import": "./dist/index.mjs",
      "require": "./dist/index.js",
      "types": "./dist/index.d.ts"
    }
  },
  "author": "Raven Satir",
  "license": "MIT"
}
