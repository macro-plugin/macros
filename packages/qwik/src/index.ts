export { qwik } from "./qwik"
export { task, vtask } from "./task"
export { resource } from "./resource"
export { store } from "./store"
export { signal } from "./signal"
export { events } from "./events"
export { computed } from "./computed"
export { link, css, scoped } from "./css"
