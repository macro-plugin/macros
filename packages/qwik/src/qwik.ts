import { ArrowFunctionExpression, CallExpression, FunctionDeclaration, FunctionExpression, LabeledStatement, VariableDeclaration } from "@swc/core"
import { WalkContext, createMacro, dummySpan } from "@macro-plugin/core"

function handleFunc (this: WalkContext, ast: FunctionDeclaration | FunctionExpression | ArrowFunctionExpression) {
  let label
  if (!ast.body || ast.body.type !== "BlockStatement") return
  for (let i = 0; i < ast.body.stmts.length; i++) {
    if (ast.body.stmts[i].type === "LabeledStatement") {
      label = ast.body.stmts[i] as LabeledStatement
      if (label.label.value === "qwik") {
        if (label.body.type === "ExpressionStatement" && label.body.expression.type === "BooleanLiteral") {
          ast.body.stmts = ast.body.stmts.filter((_, index) => index !== i)
          const isDecl = "identifier" in ast && ast.identifier != null

          if (label.body.expression.value) {
            this.import("component$", "@builder.io/qwik")

            const init = {
              type: "CallExpression",
              span: dummySpan,
              callee: {
                type: "Identifier",
                span: dummySpan,
                value: "component$",
                optional: false
              },
              arguments: [
                {
                  expression: isDecl
                    ? {
                      type: "ArrowFunctionExpression",
                      span: dummySpan,
                      params: ast.params.map(i => i.pat),
                      body: ast.body,
                      async: false,
                      generator: false,
                    }
                    : ast
                }
              ],
            } as CallExpression

            if (!isDecl) return init

            return {
              type: "VariableDeclaration",
              span: dummySpan,
              kind: "const",
              declare: false,
              declarations: [
                {
                  type: "VariableDeclarator",
                  span: dummySpan,
                  id: ast.identifier,
                  init,
                  definite: false
                }
              ]
            } as VariableDeclaration
          }
        } else {
          throw new Error("Expect an boolean in qwik macro.")
        }
      }
    }
  }
}

export const qwik = createMacro({
  FunctionDeclaration (ast) {
    return handleFunc.apply(this, [ast])
  },
  FunctionExpression (ast) {
    return handleFunc.apply(this, [ast])
  },
  ArrowFunctionExpression (ast) {
    return handleFunc.apply(this, [ast])
  }
})
