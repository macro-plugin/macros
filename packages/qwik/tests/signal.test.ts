import { signal } from "@macro-plugin/qwik"
import { transform } from "@macro-plugin/core"

test("qwik signal", () => {
  const code = `
    signal: {
      var count = 0
    }

    count += 1
  `
  expect(transform(code, { plugins: [signal] }).code).toMatchSnapshot()
})

test("qwik signal in function name", () => {
  const code = `
  signal: {
    var count = 0
  }
  function abc(count) {
    count = 3
  }
  count += 2
  `

  expect(transform(code, { plugins: [ signal ] }).code).toMatchSnapshot();
});

