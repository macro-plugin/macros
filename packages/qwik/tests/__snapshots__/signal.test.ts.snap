// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`qwik signal 1`] = `
"import { useSignal } from "@builder.io/qwik";
const count = useSignal(0);
count.value += 1;
"
`;

exports[`qwik signal in function name 1`] = `
"import { useSignal } from "@builder.io/qwik";
const count = useSignal(0);
function abc(count) {
    count = 3;
}
count.value += 2;
"
`;
